docs: introduction chapter of arc42 documentation (#156)
GenSpectrum/LAPIS: docs/arc42/01_introduction_and_goals.md

This document follows the [arc42 template](https://arc42.org/).

# Introduction and Goals

This document describes LAPIS (Lightweight API for Sequences),
which is a platform to give easy access to genomic sequence data alongside metadata of the sequenced probes.
It is used to filter potentially large sequence data and return the result to the user through web access,
so that a user can develop their own evaluation of the data.

> In the past we built https://genspectrum.org/ which involves such an API, but it is specialized for SARS-CoV-2.
> In this approach, we want to develop a generalized API that is configurable for a wider range of pathogens, especially
> viruses.

Reviewer: ```suggestion

The initial implementation of LAPIS is specialized for SARS-CoV-2 and is used by [CoV-Spectrum](https://cov-spectrum.org).
In this approach, we want to develop a generalized API that is configurable for a wider range of pathogens.
```

- Currently, I'm using "GenSpectrum" as the overall name for the organization and the whole project encompassing all dashboards, LAPIS, and other outputs.
- Striking "viruses" because maybe, it's easier than expected to include at least some bacteria instances.

SILO serves as a database with a specialized query language (SILO queries),
which could in principle be exchanged by any other database.

This document focuses on LAPIS, but we usually think of it as operated as a single unit with SILO.
We refer to this special configuration as SILO-LAPIS.

The following goals have been established for this system:

| Priority |                                                                                                             |
|----------|-------------------------------------------------------------------------------------------------------------|
| 1        | Provide an API to query large sets of genomic sequence data in a very efficient way.                        |
| 2        | Provide the infrastructure such that other researchers ("maintainers") can easily setup their own instance. |

## Requirements Overview

| Id  | Requirement                             |                                                                                        |
|-----|-----------------------------------------|----------------------------------------------------------------------------------------|

> | F1  | Create an instance for a given pathogen | Create an instance of the whole system by given configuration for different pathogens. |

Reviewer: ```suggestion
| F1  | Create an instance for a given pathogen | Create an instance of the whole system by giving a configuration for a pathogen. |
```

|                        | Ease of use          | Ease of use for maintainers to create a new instance for a new pathogen.                                          | Q:U2 |
|                        | Ease of learning     | The queries should be as easy to write as possible. We provide material to assist in learning the query language. | Q:U3 |
|                        |                      |                                                                                                                   | Q:U4 |
| Performance efficiency | Time behaviour       | It is possible to query millions of sequences in less than a second.                                              | Q:P1 |
|                        | Resource utilization | How many do we have?                                                                                              | Q:P2 |
|                        | Workload             | How many requests should be tolerable?                                                                            | Q:P3 |

> |                        | Scalability          | How many sequences do we expect?                                                                                  | Q:P4 |

Reviewer: It is hard to quantify the qualities exactly and it differs for different instances/use cases. Regarding scalability, I'd say that the "performance" (query response time, memory requirement) should grow at most linearly.

|                        |                      |                                                                                                                   |      |
| Maintainability        | Reusability          | It is possible to use LAPIS with any other database that implements the SILO query language                       | Q:M1 |
|                        | Testability          | SILO-LAPIS is well tested on end to end scope. The tests serve as examples for users and maintainers.             | Q:M2 |

[//]: # (| Cultural and Regional? |)

[//]: # (| Legal?)

[//]: # (| Security?)

[//]: # ()

[//]: # (Scenarios:)

## Stakeholders

| Role                             | Expectations                                                                     |
|----------------------------------|----------------------------------------------------------------------------------|

> | Database researcher (Alex)       | ?                                                                                |                                                                      

Reviewer: ```suggestion
| Database researcher              | Can develop new genomic data engineering algorithms for LAPIS            |                                                                      
```

> | Alignment researcher (Cornelius) | ?                                                                                |                                                                      
> | Sequence researcher (Chaoran)    | ?                                                                                |                                                                    
> | Financier (Tanja, DFG?)          | ?                                                                                |                                                                   

Reviewer: ```suggestion                                           
```

| Developer                        | Can fix bugs and add new features to software.                                   |                         

> | User                             | Can write queries to LAPIS and get a fast result.                                |                      

Reviewer: ```suggestion
| User (beginner level)   | Can write simple queries to LAPIS by hand and get a fast result.         |                      
| User (advanced level/tool developers)  | Can write advanced, possibly programmatically generated queries to LAPIS and get a fast result.                                |                      
```
